# Import Libraries and Load Dataset
The dataset is sourced from: https://www.kaggle.com/datasets/ahmedmohamed2003/retail-store-sales-dirty-for-data-cleaning/data

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r'D:\REGINA\yesgo\0 NEW\retail_store_sales.csv')

# Initial Data Analysis

## Dataset Overview

In [3]:
df.head(10)

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
5    TXN_7482416     CUST_09     Patisserie           NaN             NaN   
6    TXN_3652209     CUST_07           Food   Item_1_FOOD             5.0   
7    TXN_1372952     CUST_21      Furniture           NaN            33.5   
8    TXN_9728486     CUST_23      Furniture   Item_16_FUR            27.5   
9    TXN_2722661     CUST_25       Butchers   Item_22_BUT            36.5   

   Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      10.0        185.0  Digital Wallet    Online       2024-04-08   
1       9.0        261.0  Digital Wallet    Online       2023-07-23   
2       2.0         43.0     Credit Card    Online       2022-10-05   
3       9.0        247.5     Credit Card    Online       2022-05-07   
4       7.0         87.5  Digital Wallet    Online       2022-10-02   
5      10.0        200.0     Credit Card    Online       2023-11-30   
6       8.0         40.0     Credit Card  In-store       2023-06-10   
7       NaN          NaN  Digital Wallet  In-store       2024-04-02   
8       1.0         27.5     Credit Card  In-store       2023-04-26   
9       3.0        109.5            Cash    Online       2024-03-14   

  Discount Applied  
0             True  
1             True  
2            False  
3              NaN  
4            False  
5              NaN  
6             True  
7             True  
8            False  
9            False

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


The dataset consists of 12,575 entries with 11 columns. Here is a brief description of each column:

- Transaction ID: A unique identifier for each transaction. Always present and unique.
- Customer ID: A unique identifier for each customer. Always present and unique.
- Category: The category of the purchased item.
- Item: The name of the purchased item. This column has some missing values, with 11,362 non-null entries out of 12,575.
- Price Per Unit: The static price of a single unit of the item. This column has some missing values, with 11,966 non-null entries.
- Quantity: The quantity of the purchased item. This column has some missing values, with 11,971 non-null entries.
- Total Spent: The total amount spent on the transaction, calculated as Quantity × Price Per Unit. This column has some missing values, with 11,971 non-null entries.
- Payment Method: The method of payment used.
- Location: The location where the transaction took place.
- Transaction Date: The date of the transaction.
- Discount Applied: Whether a discount was applied (True/False). This column has some missing values, with 8,376 non-null entries.

## Summary Statistics

In [5]:
# Summary statistics for numerical variables
df.describe()

Price Per Unit      Quantity   Total Spent
count    11966.000000  11971.000000  11971.000000
mean        23.365912      5.536380    129.652577
std         10.743519      2.857883     94.750697
min          5.000000      1.000000      5.000000
25%         14.000000      3.000000     51.000000
50%         23.000000      6.000000    108.500000
75%         33.500000      8.000000    192.000000
max         41.000000     10.000000    410.000000

In [6]:
# Summary statistics for categorical variables
df.describe(include='object')

Transaction ID Customer ID   Category        Item Payment Method  \
count           12575       12575      12575       11362          12575   
unique          12575          25          8         200              3   
top       TXN_6867343     CUST_05  Furniture  Item_2_BEV           Cash   
freq                1         544       1591         126           4310   

       Location Transaction Date Discount Applied  
count     12575            12575             8376  
unique        2             1114                2  
top      Online       2022-05-30             True  
freq       6354               26             4219

In [7]:
df['Category'].unique()

array(['Patisserie', 'Milk Products', 'Butchers', 'Beverages', 'Food',
       'Furniture', 'Electric household essentials',
       'Computers and electric accessories'], dtype=object)

In [8]:
df['Payment Method'].unique()

array(['Digital Wallet', 'Credit Card', 'Cash'], dtype=object)

In [9]:
df['Location'].unique()

array(['Online', 'In-store'], dtype=object)

In [10]:
df['Discount Applied'].unique()

array([True, False, nan], dtype=object)

- Price Per Unit: Only 11,966 non-null entries, indicating missing values that need to be addressed. The average price per unit is \\$23.36, ranging from \\$5 to \\$41.
- Quantity: Only 11,971 non-null entries, meaning there are missing values to be addressed. The average quantity purchased is 5.5, ranging from 1 to 10.
- Total Spent: Only 11,971 non-null entries, indicating missing values that need to be addressed. The average spending is \\$129.65. The standard deviation suggests a spread in the data, with a minimum spending of \\$5 and a maximum of \\$410.
- Transaction ID: There are 12,575 unique transaction IDs, indicating that each transaction ID corresponds to a single item category.
- Customer ID: There are 25 unique customer IDs, suggesting repeat purchases by some customers. CUST_05 is the top customer, with 544 purchases.
- Category: There are 8 unique categories, with Furniture being the most frequently purchased category, totaling 1,591 transactions.
- Item: Only 11,362 non-null entries, meaning there are missing values to be addressed. There are 200 unique items, with Item_2_BEV being the top-selling item, purchased 126 times.
- Payment Method: There are 3 types of payment methods, which are Digital Wallet, Credit Card, and Cash. Cash is the most popular choice, used 4,310 times.
- Location: There are 2 types of transaction locations, which are Online and In-store. Online is the most popular, with 6,354 transactions.
- Transaction Date: There are 1,114 unique transaction dates, with the highest sales occurring on May 30, 2022, with 26 transactions.
- Discount Applied: Only 8,376 non-null entries, indicating missing values that need to be addressed. A total of 4,219 transactions received a discount.

# Data Cleaning and Transformation

## Handling Missing Values

In [11]:
# Calculate the percentage of missing values for each column
df.isnull().sum() / len(df) * 100

Transaction ID       0.000000
Customer ID          0.000000
Category             0.000000
Item                 9.646123
Price Per Unit       4.842942
Quantity             4.803181
Total Spent          4.803181
Payment Method       0.000000
Location             0.000000
Transaction Date     0.000000
Discount Applied    33.391650
dtype: float64

In [12]:
# Fill missing values in the Price Per Unit, Quantity, and Total Spent columns.
df['Price Per Unit'] = df['Price Per Unit'].fillna(df['Total Spent'] / df['Quantity'])
df['Total Spent'] = df['Total Spent'].fillna(df['Quantity'] * df['Price Per Unit'])
df['Quantity'] = df['Quantity'].fillna(df['Total Spent'] / df['Price Per Unit'])

In [13]:
# Recheck for any missing values
df.isnull().sum() / len(df) * 100

Transaction ID       0.000000
Customer ID          0.000000
Category             0.000000
Item                 9.646123
Price Per Unit       0.000000
Quantity             4.803181
Total Spent          4.803181
Payment Method       0.000000
Location             0.000000
Transaction Date     0.000000
Discount Applied    33.391650
dtype: float64

It is observed that the remaining missing values are in the Quantity and Total Spent columns. Since both are empty, we can simply remove the rows where both values are missing.

In [14]:
df = df.dropna(subset=['Quantity', 'Total Spent'], how='all')

The Item column is crucial for data analysis, so we can remove the 9% of missing data to ensure more accurate and meaningful insights.

In [15]:
df = df.dropna(subset=['Item'])

Since 33% of the values in the Discount Applied column are missing, we will assume that no discount was given.

In [16]:
df['Discount Applied'] = df['Discount Applied'].fillna('False')

In [17]:
# Verify the removal of missing values
df.isnull().sum().sum()

0

## Handling Duplicates

In [18]:
# Check for duplicates in the data
df.duplicated().sum()

0

It appears that there are no duplicate records in this dataset.

## Converting Data Types
The Transaction Date and Discount Applied columns are still in Object format, so they need to be converted accordingly.

In [22]:
# Check the date format of the Transaction Date column
df['Transaction Date']

0       2024-04-08
1       2023-07-23
2       2022-10-05
3       2022-05-07
4       2022-10-02
           ...    
12570   2023-09-03
12571   2022-08-12
12572   2024-08-24
12573   2023-12-30
12574   2022-08-06
Name: Transaction Date, Length: 11362, dtype: datetime64[ns]

In [23]:
df['Transaction Date'] = pd.to_datetime(df['Transaction Date'], format='%Y-%m-%d')
df['Discount Applied'] = df['Discount Applied'].astype(bool)

In [27]:
# Recheck the data types of each column
df.dtypes

Transaction ID              object
Customer ID                 object
Category                    object
Item                        object
Price Per Unit             float64
Quantity                   float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
Discount Applied              bool
dtype: object

## Validating Data Consistency

In [29]:
df['Calculated Total'] = df['Price Per Unit'] * df['Quantity']

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

   Discount Applied  Calculated Total  
0              True             185.0  
1              True             261.0  
2             False              43.0  
3              True             247.5  
4             False              87.5

# Feature Engineering

# Outlier Detection and Treatment

# Correlation Analysis